src/gax/src/retry_loop.rs

```diff
@@ -0,0 +1,187 @@
+use super::loop_state::LoopState;
+use super::retry_policy::RetryPolicy;
+use super::retry_throttler::RetryThrottler;
+use std::sync::{Arc, Mutex};
+
+/// Runs the retry loop for a given function.
+///
+/// This functions calls an inner function as long as (1) the retry policy has
+/// not expired, (2) the inner function has not returned a successful request,
+/// and (3) the retry throttler allows more calls.
+///
+/// In between calls the function waits the amount of time prescribed by the
+/// backoff policy, using `sleep` to implement any sleep.
+pub async fn retry_loop<F, S, Response>(
+    inner: F,
+    sleep: S,
+    idempotent: bool,
+    retry_throttler: Arc<Mutex<dyn RetryThrottler>>,
+    retry_policy: Arc<dyn RetryPolicy>,
+    backoff_policy: Arc<dyn BackoffPolicy>,
+) -> Result<Response>
+where
+    F: AsyncFn(Option<std::time::Duration>) -> Result<Response> + Send,
+    S: AsyncFn(std::time::Duration) -> () + Send,
+{
+    let loop_start = std::time::Instant::now();
+    let mut attempt_count = 0;
+    loop {
+        let remaining_time = retry_policy.remaining_time(loop_start, attempt_count);
+        let throttle = if attempt_count == 0 {
+            false
+        } else {
+            let t = retry_throttler
+                .lock()
+                .expect("retry throttler lock is poisoned");
+            t.throttle_retry_attempt()
+        };
+        if throttle {
+            // This counts as an error for the purposes of the retry policy.
+            if let Some(error) = retry_policy.on_throttle(loop_start, attempt_count) {
+                return Err(error);
+            }
+            let delay = backoff_policy.on_failure(loop_start, attempt_count);
+            sleep(delay).await;
+            continue;
+        }
+        attempt_count += 1;
+        match inner(remaining_time).await {
+            Ok(r) => {
+                retry_throttler
+                    .lock()
+                    .expect("retry throttler lock is poisoned")
+                    .on_success();
+                return Ok(r);
+            }
+            Err(e) => {
+                let flow = retry_policy.on_error(loop_start, attempt_count, idempotent, e);
+                let delay = backoff_policy.on_failure(loop_start, attempt_count);
+                retry_throttler
+                    .lock()
+                    .expect("retry throttler lock is poisoned")
+                    .on_retry_failure(&flow);
+                on_error(&sleep, flow, delay).await?;
+            }
+        };
+    }
+}
+
+async fn on_error<B>(
+    backoff: &B,
+    retry_flow: LoopState,
+    backoff_delay: std::time::Duration,
+) -> Result<()>
+where
+    B: AsyncFn(std::time::Duration) -> (),
+{
+    match retry_flow {
+        LoopState::Permanent(e) | LoopState::Exhausted(e) => Err(e),
+        LoopState::Continue(_e) => {
+            backoff(backoff_delay).await;
+            Ok(())
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
comment: At first, I thought that the tests should move to this interface, but I think its fine that the tests are in `src/gax-internal/tests/retry_loop.rs`

Reply:
I think we should refactor the tests, but we can do that later.  For now, we have good coverage and the initial test is a good start.